assistant/assistant.go

```diff
@@ -0,0 +1,186 @@
+package assistant
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/influxdata/telegraf/agent"
+	"github.com/influxdata/telegraf/internal"
+)
+
+/*
+Assistant is a client to facilitate communications between Agent and Cloud.
+*/
+type Assistant struct {
+	config     *AssistantConfig // stores plugins
+	connection *websocket.Conn  // Active websocket connection
+	ctx        context.Context  // go's context
+	done       chan bool        // Channel used to stop server listener
+	agent      *agent.Agent     // Pointer to agent to issue commands
+}
+
+/*
+AssistantConfig allows us to configure where to connect to and other params
+for the agent.
+*/
+type AssistantConfig struct {
+	Host          string
+	Path          string
+	RetryInterval int
+}
+
+func (astConfig *AssistantConfig) fillDefaults() {
+	if astConfig.Host == "" {
+		astConfig.Host = "localhost:8080"
+	}
+	if astConfig.Path == "" {
+		astConfig.Path = "/echo"
+	}
+	if astConfig.RetryInterval == 0 {
+		astConfig.RetryInterval = 15
+	}
+}
+
+// NewAssistant returns an Assistant for the given Config.
+func NewAssistant(ctx context.Context, config *AssistantConfig, agent *agent.Agent) (*Assistant, error) {
```

Comment:
nit: Normally in Golang a function called ```NewXXX``` just returns the struct (but doesn't really start it per se). This function should probably be called StartAssistant or something, because it's actually starting the connection and loop and everything.